src/main/java/appeng/fluids/parts/FluidHandlerAdapter.java

```diff
@@ -0,0 +1,301 @@
+
+import appeng.api.AEApi;
+import appeng.api.config.Actionable;
+import appeng.api.networking.security.IActionSource;
+import appeng.api.networking.storage.IBaseMonitor;
+import appeng.api.networking.ticking.TickRateModulation;
+import appeng.api.storage.IMEInventory;
+import appeng.api.storage.IMEMonitorHandlerReceiver;
+import appeng.api.storage.IStorageChannel;
+import appeng.api.storage.channels.IFluidStorageChannel;
+import appeng.api.storage.data.IAEFluidStack;
+import appeng.api.storage.data.IItemList;
+import appeng.me.GridAccessException;
+import appeng.me.helpers.IGridProxyable;
+import appeng.me.storage.ITickingMonitor;
+import appeng.util.item.AEFluidStack;
+
+
+/**
+ * Wraps an Fluid Handler in such a way that it can be used as an IMEInventory for fluids.
+ *
+ * @author BrockWS
+ * @version rv6 - 22/05/2018
+ * @since rv6 22/05/2018
+ */
+public class FluidHandlerAdapter implements IMEInventory<IAEFluidStack>, IBaseMonitor<IAEFluidStack>, ITickingMonitor
+{
+	private final Map<IMEMonitorHandlerReceiver<IAEFluidStack>, Object> listeners = new HashMap<>();
+	private IActionSource source;
+	private final IFluidHandler fluidHandler;
+	private final IGridProxyable proxyable;
+	private final FluidHandlerAdapter.InventoryCache cache;
+
+	FluidHandlerAdapter( IFluidHandler fluidHandler, IGridProxyable proxy )
+	{
+		this.fluidHandler = fluidHandler;
+		this.proxyable = proxy;
+		this.cache = new FluidHandlerAdapter.InventoryCache( this.fluidHandler );
+	}
+
+	@Override
+	public IAEFluidStack injectItems( IAEFluidStack input, Actionable type, IActionSource src )
+	{
+		FluidStack fluidStack = input.getFluidStack();
+
+		// Insert
+		// TODO: May need to brute force like ItemHandlerAdapter
+		int wasFillled = this.fluidHandler.fill( fluidStack, type != Actionable.SIMULATE );
+		int remaining = fluidStack.amount - wasFillled;
+		if( fluidStack.amount == remaining )
+		{
+			// The stack was unmodified, target tank is full
+			return input;
+		}
+
+		if( type == Actionable.MODULATE )
+		{
+			try
+			{
+				this.proxyable.getProxy().getTick().alertDevice( this.proxyable.getProxy().getNode() );
+			}
+			catch( GridAccessException ex )
+			{
+				// meh
+			}
+		}
+
+		fluidStack.amount = remaining;
+
+		return AEFluidStack.fromFluidStack( fluidStack );
+	}
+
+	@Override
+	public IAEFluidStack extractItems( IAEFluidStack request, Actionable mode, IActionSource src )
+	{
+		FluidStack requestedFluidStack = request.getFluidStack();
+		final boolean doDrain = ( mode == Actionable.MODULATE );
+
+		// Drain the fluid from the tank
+		FluidStack gathered = this.fluidHandler.drain( requestedFluidStack, doDrain );
+		if( gathered == null )
+		{
+			// If nothing was pulled from the tank, return null
+			return null;
+		}
+
+		if( mode == Actionable.MODULATE )
+		{
+			try
+			{
+				this.proxyable.getProxy().getTick().alertDevice( this.proxyable.getProxy().getNode() );
+			}
+			catch( GridAccessException ex )
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1166)